File: GitReader/Structures/RepositoryExtension.cs
////////////////////////////////////////////////////////////////////////////
//
// GitReader - Lightweight Git local repository traversal library.
// Copyright (c) Kouji Matsui (@kozy_kekyo, @kekyo@mi.kekyo.net)
//
// Licensed under Apache-v2: https://opensource.org/licenses/Apache-2.0
//
////////////////////////////////////////////////////////////////////////////

using GitReader.Collections;
using GitReader.Internal;
using GitReader.Structures;
using System;
using System.IO;
using System.Threading;
using System.Threading.Tasks;

namespace GitReader.Structures;

/// <summary>
/// Provides extension methods for structured repository operations.
/// </summary>
public static class RepositoryExtension
{
    /// <summary>
    /// Gets a commit from the repository by its hash.
    /// </summary>
    /// <param name="repository">The structured repository.</param>
    /// <param name="commit">The commit hash to retrieve.</param>
    /// <param name="ct">The cancellation token.</param>
    /// <returns>A task that returns the commit, or null if not found.</returns>
    public static Task<Commit?> GetCommitAsync(
        this StructuredRepository repository,
        Hash commit, CancellationToken ct = default) =>
        StructuredRepositoryFacade.GetCommitDirectlyAsync(repository, commit, ct);

    /// <summary>
    /// Gets the head commit of the specified branch.
    /// </summary>
    /// <param name="branch">The branch to get the head commit from.</param>
    /// <param name="ct">The cancellation token.</param>
    /// <returns>A task that returns the head commit of the branch.</returns>
    public static Task<Commit> GetHeadCommitAsync(
        this Branch branch,
        CancellationToken ct = default) =>
        StructuredRepositoryFacade.GetCommitAsync(branch, ct);

    /// <summary>
    /// Gets the commit that the specified tag points to.
    /// </summary>
    /// <param name="tag">The tag to get the commit from.</param>
    /// <param name="ct">The cancellation token.</param>
    /// <returns>A task that returns the commit that the tag points to.</returns>
    /// <exception cref="InvalidOperationException">Thrown when the tag does not point to a commit.</exception>
    public static Task<Commit> GetCommitAsync(
        this Tag tag,
        CancellationToken ct = default) =>
        tag.Type switch
        {
            ObjectTypes.Commit => StructuredRepositoryFacade.GetCommitAsync(tag, ct),
            _ => throw new InvalidOperationException($"Could not get commit: Type={tag.Type}"),
        };

    /// <summary>
    /// Gets the annotation associated with the specified tag.
    /// </summary>
    /// <param name="tag">The tag to get the annotation from.</param>
    /// <param name="ct">The cancellation token.</param>
    /// <returns>A task that returns the annotation of the tag.</returns>
    public static Task<Annotation> GetAnnotationAsync(
        this Tag tag,
        CancellationToken ct = default) =>
        StructuredRepositoryFacade.GetAnnotationAsync(tag, ct);

    /// <summary>
    /// Gets the commit associated with the specified stash entry.
    /// </summary>
    /// <param name="stash">The stash entry to get the commit from.</param>
    /// <param name="ct">The cancellation token.</param>
    /// <returns>A task that returns the commit associated with the stash.</returns>
    public static Task<Commit> GetCommitAsync(
        this Stash stash,
        CancellationToken ct = default) =>
        StructuredRepositoryFacade.GetCommitAsync(stash, ct);

    /// <summary>
    /// Gets the current commit from the specified reflog entry.
    /// </summary>
    /// <param name="reflog">The reflog entry.</param>
    /// <param name="ct">The cancellation token.</param>
    /// <returns>A task that returns the current commit of the reflog entry.</returns>
    public static Task<Commit> GetCurrentCommitAsync(
        this ReflogEntry reflog,
        CancellationToken ct = default) =>
        StructuredRepositoryFacade.GetCommitAsync(reflog, reflog.Commit, ct);

    /// <summary>
    /// Gets the old commit from the specified reflog entry.
    /// </summary>
    /// <param name="reflog">The reflog entry.</param>
    /// <param name="ct">The cancellation token.</param>
    /// <returns>A task that returns the old commit of the reflog entry.</returns>
    public static Task<Commit> GetOldCommitAsync(
        this ReflogEntry reflog,
        CancellationToken ct = default) =>
        StructuredRepositoryFacade.GetCommitAsync(reflog, reflog.OldCommit, ct);

    /// <summary>
    /// Gets the primary parent commit of the specified commit.
    /// </summary>
    /// <param name="commit">The commit to get the primary parent from.</param>
    /// <param name="ct">The cancellation token.</param>
    /// <returns>A task that returns the primary parent commit, or null if it has no parents.</returns>
    public static Task<Commit?> GetPrimaryParentCommitAsync(
        this Commit commit,
        CancellationToken ct = default) =>
        StructuredRepositoryFacade.GetPrimaryParentAsync(commit, ct);

    /// <summary>
    /// Gets all parent commits of the specified commit.
    /// </summary>
    /// <param name="commit">The commit to get the parents from.</param>
    /// <param name="ct">The cancellation token.</param>
    /// <returns>A task that returns an array of parent commits.</returns>
    public static Task<Commit[]> GetParentCommitsAsync(
        this Commit commit,
        CancellationToken ct = default) =>
        StructuredRepositoryFacade.GetParentsAsync(commit, ct);

    /// <summary>
    /// Gets the tree root associated with the specified commit.
    /// </summary>
    /// <param name="commit">The commit to get the tree root from.</param>
    /// <param name="ct">The cancellation token.</param>
    /// <returns>A task that returns the tree root of the commit.</returns>
    public static Task<TreeRoot> GetTreeRootAsync(
        this Commit commit,
        CancellationToken ct = default) =>
        StructuredRepositoryFacade.GetTreeAsync(commit, ct);

    /// <summary>
    /// Opens a blob stream for the specified tree blob entry.
    /// </summary>
    /// <param name="entry">The tree blob entry to open.</param>
    /// <param name="ct">The cancellation token.</param>
    /// <returns>A task that returns a stream for reading the blob content.</returns>
    public static Task<Stream> OpenBlobAsync(
        this TreeBlobEntry entry,
        CancellationToken ct = default) =>
        StructuredRepositoryFacade.OpenBlobAsync(entry, ct);

    /// <summary>
    /// Opens the submodule repository associated with the specified tree submodule entry.
    /// </summary>
    /// <param name="subModule">The tree submodule entry.</param>
    /// <param name="ct">The cancellation token.</param>
    /// <returns>A task that returns the structured repository of the submodule.</returns>
    public static Task<StructuredRepository> OpenSubModuleAsync(
        this TreeSubModuleEntry subModule,
        CancellationToken ct = default) =>
        StructuredRepositoryFacade.OpenSubModuleAsync(subModule, ct);

    /// <summary>
    /// Gets the full commit message of the specified commit.
    /// </summary>
    /// <param name="commit">The commit to get the message from.</param>
    /// <returns>The full commit message.</returns>
    public static string GetMessage(
        this Commit commit) =>
        commit.message;
    
    /// <summary>
    /// Gets the head reflog entries for the repository.
    /// </summary>
    /// <param name="repository">The structured repository.</param>
    /// <param name="ct">The cancellation token.</param>
    /// <returns>A task that returns an array of reflog entries.</returns>
    public static Task<ReflogEntry[]> GetHeadReflogsAsync(
        this StructuredRepository repository, CancellationToken ct = default) =>
        StructuredRepositoryFacade.GetHeadReflogsAsync(
            repository, new WeakReference(repository), ct);

    /// <summary>
    /// Gets working directory status.
    /// </summary>
    /// <param name="repository">The repository to get working directory status from.</param>
    /// <param name="ct">The cancellation token.</param>
    /// <returns>A Task containing the structured working directory status.</returns>
    public static Task<WorkingDirectoryStatus> GetWorkingDirectoryStatusAsync(
        this StructuredRepository repository,
        CancellationToken ct = default) =>
#if NET45_OR_GREATER || NETSTANDARD || NETCOREAPP
        StructuredRepositoryFacade.GetWorkingDirectoryStatusAsync(
            repository, Internal.Glob.nothingFilter, ct).AsTask();
#else
        StructuredRepositoryFacade.GetWorkingDirectoryStatusAsync(
            repository, Internal.Glob.nothingFilter, ct);
#endif

    /// <summary>
    /// Gets working directory status with an override glob filter.
    /// </summary>
    /// <param name="repository">The repository to get working directory status from.</param>
    /// <param name="overrideGlobFilter">A predicate override glob filter function.</param>
    /// <param name="ct">The cancellation token.</param>
    /// <returns>A Task containing the structured working directory status.</returns>
    public static Task<WorkingDirectoryStatus> GetWorkingDirectoryStatusAsync(
        this StructuredRepository repository,
        GlobFilter overrideGlobFilter,
        CancellationToken ct = default) =>
#if NET45_OR_GREATER || NETSTANDARD || NETCOREAPP
        StructuredRepositoryFacade.GetWorkingDirectoryStatusAsync(
            repository, overrideGlobFilter, ct).AsTask();
#else
        StructuredRepositoryFacade.GetWorkingDirectoryStatusAsync(
            repository, overrideGlobFilter, ct);
#endif

    /// <summary>
    /// Gets all worktrees associated with the repository.
    /// </summary>
    /// <param name="repository">The structured repository.</param>
    /// <param name="ct">The cancellation token.</param>
    /// <returns>A task that returns a read-only array of worktrees.</returns>
    public static Task<ReadOnlyArray<Worktree>> GetWorktreesAsync(
        this StructuredRepository repository, CancellationToken ct = default) =>
#if NET45_OR_GREATER || NETSTANDARD || NETCOREAPP
        WorktreeAccessor.GetStructuredWorktreesAsync(repository, ct).AsTask();
#else
        WorktreeAccessor.GetStructuredWorktreesAsync(repository, ct);
#endif

    /// <summary>
    /// Deconstructs a StructuredRepository into its component parts (including all branches).
    /// </summary>
    /// <param name="repository">The repository to deconstruct.</param>
    /// <param name="gitPath">The Git repository path.</param>
    /// <param name="remoteUrls">The remote URLs configured for the repository.</param>
    /// <param name="head">The current head branch, or null if in detached HEAD state.</param>
    /// <param name="branchesAll">All branches grouped by name (including remote branches).</param>
    /// <param name="tags">All tags in the repository.</param>
    public static void Deconstruct(
        this StructuredRepository repository,
        out string gitPath,
        out ReadOnlyDictionary<string, string> remoteUrls,
        out Branch? head,
        out ReadOnlyDictionary<string, Branch[]> branchesAll,
        out ReadOnlyDictionary<string, Tag> tags)
    {
        gitPath = repository.GitPath;
        remoteUrls = repository.RemoteUrls;
        head = repository.head;
        branchesAll = repository.branchesAll;
        tags = repository.tags;
    }

    /// <summary>
    /// Deconstructs a StructuredRepository into its component parts (primary branches only).
    /// </summary>
    /// <param name="repository">The repository to deconstruct.</param>
    /// <param name="gitPath">The Git repository path.</param>
    /// <param name="remoteUrls">The remote URLs configured for the repository.</param>
    /// <param name="head">The current head branch, or null if in detached HEAD state.</param>
    /// <param name="branches">The primary branches in the repository.</param>
    /// <param name="tags">All tags in the repository.</param>
    public static void Deconstruct(
        this StructuredRepository repository,
        out string gitPath,
        out ReadOnlyDictionary<string, string> remoteUrls,
        out Branch? head,
        out ReadOnlyDictionary<string, Branch> branches,
        out ReadOnlyDictionary<string, Tag> tags)
    {
        gitPath = repository.GitPath;
        remoteUrls = repository.RemoteUrls;
        head = repository.head;
        branches = repository.Branches;
        tags = repository.Tags;
    }

    /// <summary>
    /// Deconstructs a Branch into its component parts.
    /// </summary>
    /// <param name="branch">The branch to deconstruct.</param>
    /// <param name="name">The name of the branch.</param>
    /// <param name="head">The head commit hash of the branch.</param>
    public static void Deconstruct(
        this Branch branch,
        out string name,
        out Hash head)
    {
        name = branch.Name;
        head = branch.Head;
    }

    /// <summary>
    /// Deconstructs a Commit into its component parts (with separate subject and body).
    /// </summary>
    /// <param name="commit">The commit to deconstruct.</param>
    /// <param name="hash">The commit hash.</param>
    /// <param name="author">The author signature.</param>
    /// <param name="committer">The committer signature.</param>
    /// <param name="subject">The commit subject (first line of the message).</param>
    /// <param name="body">The commit body (remaining lines of the message).</param>
    /// <remarks>
    /// The subject line is the first line of the commit message. It is nearly as git command `git log --format=%s`.
    /// The body is the rest of the commit message after the first blank line. It is nearly as git command `git log --format=%b`.
    /// </remarks>
    public static void Deconstruct(
        this Commit commit,
        out Hash hash,
        out Signature author,
        out Signature committer,
        out string subject,
        out string body)
    {
        hash = commit.Hash;
        author = commit.Author;
        committer = commit.Committer;
        subject = commit.Subject;
        body = commit.Body;
    }

    /// <summary>
    /// Deconstructs a Commit into its component parts (with full message).
    /// </summary>
    /// <param name="commit">The commit to deconstruct.</param>
    /// <param name="hash">The commit hash.</param>
    /// <param name="author">The author signature.</param>
    /// <param name="committer">The committer signature.</param>
    /// <param name="message">The full commit message.</param>
    public static void Deconstruct(
        this Commit commit,
        out Hash hash,
        out Signature author,
        out Signature committer,
        out string message)
    {
        hash = commit.Hash;
        author = commit.Author;
        committer = commit.Committer;
        message = commit.message;
    }

    /// <summary>
    /// Deconstructs a Tag into its component parts.
    /// </summary>
    /// <param name="tag">The tag to deconstruct.</param>
    /// <param name="tagHash">The tag hash.</param>
    /// <param name="type">The type of the tag.</param>
    /// <param name="objectHash">The hash of the object the tag points to.</param>
    /// <param name="name">The name of the tag.</param>
    public static void Deconstruct(
        this Tag tag,
        out Hash? tagHash,
        out ObjectTypes type,
        out Hash objectHash,
        out string name)
    {
        tagHash = tag.TagHash;
        type = tag.Type;
        objectHash = tag.ObjectHash;
        name = tag.Name;
    }

    /// <summary>
    /// Deconstructs an Annotation into its component parts.
    /// </summary>
    /// <param name="annotation">The annotation to deconstruct.</param>
    /// <param name="tagger">The tagger signature.</param>
    /// <param name="message">The annotation message.</param>
    public static void Deconstruct(
        this Annotation annotation,
        out Signature? tagger,
        out string? message)
    {
        tagger = annotation.Tagger;
        message = annotation.Message;
    }

    /// <summary>
    /// Deconstructs a WorkingDirectoryFile into its component parts.
    /// </summary>
    /// <param name="file">The working directory file to deconstruct.</param>
    /// <param name="path">The path of the file.</param>
    /// <param name="status">The status of the file.</param>
    /// <param name="indexHash">The hash of the file in the index.</param>
    /// <param name="workingTreeHash">The hash of the file in the working tree.</param>
    public static void Deconstruct(
        this WorkingDirectoryFile file,
        out string path,
        out FileStatus status,
        out Hash? indexHash,
        out Hash? workingTreeHash)
    {
        path = file.Path;
        status = file.Status;
        indexHash = file.IndexHash;
        workingTreeHash = file.WorkingTreeHash;
    }

    /// <summary>
    /// Deconstructs a WorkingDirectoryStatus into its component parts.
    /// </summary>
    /// <param name="status">The working directory status to deconstruct.</param>
    /// <param name="stagedFiles">The staged files.</param>
    /// <param name="unstagedFiles">The unstaged files.</param>
    /// <param name="untrackedFiles">The untracked files.</param>
    public static void Deconstruct(
        this WorkingDirectoryStatus status,
        out ReadOnlyArray<WorkingDirectoryFile> stagedFiles,
        out ReadOnlyArray<WorkingDirectoryFile> unstagedFiles,
        out ReadOnlyArray<WorkingDirectoryFile> untrackedFiles)
    {
        stagedFiles = status.StagedFiles;
        unstagedFiles = status.UnstagedFiles;
        untrackedFiles = status.UntrackedFiles;
    }

    /// <summary>
    /// Deconstructs a Worktree into its component parts.
    /// </summary>
    /// <param name="worktree">The worktree to deconstruct.</param>
    /// <param name="name">The name of the worktree.</param>
    /// <param name="path">The path of the worktree.</param>
    /// <param name="head">The head commit hash of the worktree.</param>
    /// <param name="branch">The branch of the worktree.</param>
    /// <param name="status">The status of the worktree.</param>
    public static void Deconstruct(
        this Worktree worktree,
        out string name,
        out string path,
        out Hash? head,
        out string? branch,
        out WorktreeStatus status)
    {
        name = worktree.Name;
        path = worktree.Path;
        head = worktree.Head;
        branch = worktree.Branch;
        status = worktree.Status;
    }
}

File: FSharp.GitReader/Structures/RepositoryExtension.fs
////////////////////////////////////////////////////////////////////////////
//
// GitReader - Lightweight Git local repository traversal library.
// Copyright (c) Kouji Matsui (@kozy_kekyo, @kekyo@mi.kekyo.net)
//
// Licensed under Apache-v2: https://opensource.org/licenses/Apache-2.0
//
////////////////////////////////////////////////////////////////////////////

namespace GitReader.Structures

open GitReader
open GitReader.Internal
open GitReader.Structures
open System
open System.Threading
open System.Threading.Tasks

/// <summary>
/// Provides F#-specific extension methods for structured repository operations.
/// </summary>
[<AutoOpen>]
module public RepositoryExtension =

    type StructuredRepository with
        /// <summary>
        /// Gets the current HEAD as an optional value.
        /// </summary>
        /// <returns>Some HEAD reference if available; otherwise, None.</returns>
        member repository.getCurrentHead() =
            match repository.Head with
            | null -> None
            | _ -> Some repository.Head

        /// <summary>
        /// Gets the commit object for the specified hash.
        /// </summary>
        /// <param name="commit">The hash of the commit.</param>
        /// <param name="ct">Optional cancellation token.</param>
        /// <returns>An async computation that returns an optional commit object.</returns>
        member repository.getCommit(commit: Hash, ?ct: CancellationToken) = async {
            let! c = StructuredRepositoryFacade.GetCommitDirectlyAsync(
                repository, commit, unwrapCT ct).asAsync()
            return match c with
                   | null -> None
                   | _ -> Some c
        }

        /// <summary>
        /// Gets the reflog entries for the HEAD.
        /// </summary>
        /// <param name="ct">Optional cancellation token.</param>
        /// <returns>An async computation that returns HEAD reflog entries.</returns>
        member repository.getHeadReflogs(?ct: CancellationToken) =
            StructuredRepositoryFacade.GetHeadReflogsAsync(
                repository, WeakReference(repository), unwrapCT ct).asAsync()

        /// <summary>
        /// Gets the working directory status.
        /// </summary>
        /// <param name="ct">Optional cancellation token.</param>
        /// <returns>An async computation that returns the working directory status.</returns>
        member repository.getWorkingDirectoryStatus(?ct: CancellationToken) =
            StructuredRepositoryFacade.GetWorkingDirectoryStatusAsync(
                repository, Glob.nothingFilter, unwrapCT ct).asAsync()

        /// <summary>
        /// Gets the working directory status with an override path filter.
        /// </summary>
        /// <param name="overrideGlobFilter">A predicate override glob filter function.</param>
        /// <param name="ct">Optional cancellation token.</param>
        /// <returns>An async computation that returns the filtered working directory status.</returns>
        member repository.getWorkingDirectoryStatus(
            overrideGlobFilter: GlobFilter, ?ct: CancellationToken) =
            StructuredRepositoryFacade.GetWorkingDirectoryStatusAsync(
                repository, overrideGlobFilter, unwrapCT ct).asAsync()

        /// <summary>
        /// Gets all worktrees associated with the repository.
        /// </summary>
        /// <param name="ct">Optional cancellation token.</param>
        /// <returns>An async computation that returns all worktrees.</returns>
        member repository.getWorktrees(?ct: CancellationToken) =
            WorktreeAccessor.GetStructuredWorktreesAsync(
                repository, unwrapCT ct).asAsync()

    type Branch with
        /// <summary>
        /// Gets the HEAD commit of the branch.
        /// </summary>
        /// <param name="ct">Optional cancellation token.</param>
        /// <returns>An async computation that returns the HEAD commit.</returns>
        member branch.getHeadCommit(?ct: CancellationToken) =
            StructuredRepositoryFacade.GetCommitAsync(
                branch, unwrapCT ct).asAsync()

    type Commit with
        /// <summary>
        /// Gets the primary parent commit, if available.
        /// </summary>
        /// <param name="ct">Optional cancellation token.</param>
        /// <returns>An async computation that returns an optional primary parent commit.</returns>
        member commit.getPrimaryParentCommit(?ct: CancellationToken) = async {
            let! c = StructuredRepositoryFacade.GetPrimaryParentAsync(
                commit, unwrapCT ct).asAsync()
            return match c with
                   | null -> None
                   | _ -> Some c
        }

        /// <summary>
        /// Gets all parent commits.
        /// </summary>
        /// <param name="ct">Optional cancellation token.</param>
        /// <returns>An async computation that returns all parent commits.</returns>
        member commit.getParentCommits(?ct: CancellationToken) =
            StructuredRepositoryFacade.GetParentsAsync(
                commit, unwrapCT ct).asAsync()
                
        /// <summary>
        /// Gets the tree root of the commit.
        /// </summary>
        /// <param name="ct">Optional cancellation token.</param>
        /// <returns>An async computation that returns the tree root.</returns>
        member commit.getTreeRoot(?ct: CancellationToken) =
            StructuredRepositoryFacade.GetTreeAsync(
                commit, unwrapCT ct).asAsync()
                
        /// <summary>
        /// Gets the commit message.
        /// </summary>
        /// <returns>The commit message.</returns>
        member commit.getMessage() =
            commit.message
            
    type Tag with
        /// <summary>
        /// Gets the commit referenced by the tag.
        /// </summary>
        /// <param name="ct">Optional cancellation token.</param>
        /// <returns>An async computation that returns the commit.</returns>
        /// <exception cref="InvalidOperationException">Thrown when the tag type is not a commit.</exception>
        member tag.getCommit(?ct: CancellationToken) =
            match tag.Type with
            | ObjectTypes.Commit -> StructuredRepositoryFacade.GetCommitAsync(tag, unwrapCT ct).asAsync()
            | _ -> raise (InvalidOperationException $"Could not get commit: Type={tag.Type}")
            
        /// <summary>
        /// Gets the annotation of the tag.
        /// </summary>
        /// <param name="ct">Optional cancellation token.</param>
        /// <returns>An async computation that returns the tag annotation.</returns>
        member tag.getAnnotation(?ct: CancellationToken) =
            StructuredRepositoryFacade.GetAnnotationAsync(tag, unwrapCT ct).asAsync()
       
    type Stash with
        /// <summary>
        /// Gets the commit associated with the stash.
        /// </summary>
        /// <param name="ct">Optional cancellation token.</param>
        /// <returns>An async computation that returns the stash commit.</returns>
        member stash.getCommit(?ct: CancellationToken) =
            StructuredRepositoryFacade.GetCommitAsync(
                stash, unwrapCT ct).asAsync()
            
    type ReflogEntry with
        /// <summary>
        /// Gets the current commit of the reflog entry.
        /// </summary>
        /// <param name="ct">Optional cancellation token.</param>
        /// <returns>An async computation that returns the current commit.</returns>
        member reflog.getCurrentCommit(?ct: CancellationToken) =
            StructuredRepositoryFacade.GetCommitAsync(
                reflog, reflog.Commit, unwrapCT ct).asAsync()
                
        /// <summary>
        /// Gets the old commit of the reflog entry.
        /// </summary>
        /// <param name="ct">Optional cancellation token.</param>
        /// <returns>An async computation that returns the old commit.</returns>
        member reflog.getOldCommit(?ct: CancellationToken) =
            StructuredRepositoryFacade.GetCommitAsync(
                reflog, reflog.OldCommit, unwrapCT ct).asAsync()

    type TreeBlobEntry with
        /// <summary>
        /// Opens the blob content stream.
        /// </summary>
        /// <param name="ct">Optional cancellation token.</param>
        /// <returns>An async computation that returns the blob stream.</returns>
        member entry.openBlob(?ct: CancellationToken) =
            StructuredRepositoryFacade.OpenBlobAsync(
                entry, unwrapCT ct).asAsync()

    type TreeSubModuleEntry with
        /// <summary>
        /// Opens the submodule repository.
        /// </summary>
        /// <param name="ct">Optional cancellation token.</param>
        /// <returns>An async computation that returns the submodule repository.</returns>
        member entry.openSubModule(?ct: CancellationToken) =
            StructuredRepositoryFacade.OpenSubModuleAsync(
                entry, unwrapCT ct).asAsync()

    /// <summary>
    /// Active pattern for deconstructing a StructuredRepository into its component parts.
    /// </summary>
    /// <param name="repository">The repository to deconstruct.</param>
    /// <returns>A tuple containing the Git path, remote URLs, optional HEAD, all branches, and tags.</returns>
    let (|StructuredRepository|) (repository: StructuredRepository) =
        (repository.GitPath,
         repository.RemoteUrls,
         (match repository.head with
          | null -> None
          | _ -> Some repository.head),
         repository.branchesAll,
         repository.tags)

    /// <summary>
    /// Active pattern for deconstructing a Branch into its component parts.
    /// </summary>
    /// <param name="branch">The branch to deconstruct.</param>
    /// <returns>A tuple containing the name and HEAD hash.</returns>
    let (|Branch|) (branch: Branch) =
        (branch.Name, branch.Head)

    /// <summary>
    /// Active pattern for deconstructing a Commit into its component parts.
    /// </summary>
    /// <param name="commit">The commit to deconstruct.</param>
    /// <returns>A tuple containing the hash, author, committer, subject, and body.</returns>
    let (|Commit|) (commit: Commit) =
        (commit.Hash, commit.Author, commit.Committer, commit.Subject, commit.Body)

    /// <summary>
    /// Active pattern for deconstructing a Tag into its component parts.
    /// </summary>
    /// <param name="tag">The tag to deconstruct.</param>
    /// <returns>A tuple containing the optional tag hash, type, object hash, and name.</returns>
    let (|Tag|) (tag: Tag) =
        ((match tag.TagHash.HasValue with
         | false -> None
         | _ -> Some tag.TagHash.Value),
         tag.Type, tag.ObjectHash, tag.Name)

    /// <summary>
    /// Active pattern for deconstructing an Annotation into its component parts.
    /// </summary>
    /// <param name="annotation">The annotation to deconstruct.</param>
    /// <returns>A tuple containing the optional tagger and optional message.</returns>
    let (|Annotation|) (annotation: Annotation) =
        ((match annotation.Tagger.HasValue with
          | false -> None
          | _ -> Some annotation.Tagger.Value),
         (match annotation.Message with
          | null -> None
          | _ -> Some annotation.Message))

    /// <summary>
    /// Active pattern for deconstructing a WorkingDirectoryFile into its component parts.
    /// </summary>
    /// <param name="file">The working directory file to deconstruct.</param>
    /// <returns>A tuple containing the path, status, optional index hash, and optional working tree hash.</returns>
    let (|WorkingDirectoryFile|) (file: WorkingDirectoryFile) =
        (file.Path, file.Status, file.IndexHash |> wrapOptionV, file.WorkingTreeHash |> wrapOptionV)

    /// <summary>
    /// Active pattern for deconstructing a WorkingDirectoryStatus into its component parts.
    /// </summary>
    /// <param name="status">The working directory status to deconstruct.</param>
    /// <returns>A tuple containing staged files, unstaged files, and untracked files.</returns>
    let (|WorkingDirectoryStatus|) (status: WorkingDirectoryStatus) =
        (status.StagedFiles, status.UnstagedFiles, status.UntrackedFiles)

    /// <summary>
    /// Active pattern for deconstructing a Worktree into its component parts.
    /// </summary>
    /// <param name="worktree">The worktree to deconstruct.</param>
    /// <returns>A tuple containing the name, path, optional HEAD, optional branch, and status.</returns>
    let (|Worktree|) (worktree: Worktree) =
        (worktree.Name, worktree.Path, worktree.Head |> wrapOptionV, 
         (match worktree.Branch with
          | null -> None
          | _ -> Some worktree.Branch), worktree.Status)
